### PR TITLE
feat(metrics_indexer): Wire up the writes limiter to postgres [INGEST-1380]

### PR DESCRIPTION
**Pull request by untitaker**
Split from https://github.com/getsentry/sentry/pull/36531

**Review https://github.com/getsentry/sentry/pull/37173 first!**

This wires up all the new code we added together so that the sentry option added in https://github.com/getsentry/sentry/pull/37173 can actually be used in production.

### CODE REVIEW

File: src/sentry/sentry_metrics/indexer/postgres_v2.py [hunk 1]
```diff
@@ -136,6 +137,12 @@ def bulk_record(
             indexer_cache.set_many(new_results_to_cache, use_case_id.value)
             return cache_key_results.merge(db_read_key_results)
 
+        (
+            write_limits_state,
+            db_write_keys,
```
Comment (onewland) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:142]: is it possible to not reassign the `db_write_keys` variable and create a new one here? I could see that causing confusion

**Review by onewland — COMMENTED** [review]
I'm good with this generally, but I think we should consider not reassign-ing `db_write_keys` where I pointed it out above

**Review by onewland — APPROVED** [review]

File: src/sentry/sentry_metrics/indexer/postgres_v2.py [hunk 2]
```diff
@@ -136,24 +137,39 @@ def bulk_record(
...
-            organization_id, string = write_pair
-            new_records.append(
-                self._table(use_case_id)(organization_id=int(organization_id), string=string)
-            )
+        with writes_limiter.check_write_limits(use_case_id, db_write_keys) as writes_limiter_state:
+            # After the DB has successfully committed writes, we exit this
+            # context manager and consume quotas. If the DB crashes we
+            # shouldn't consume quota.
+            filtered_db_write_keys = writes_limiter_state.accepted_keys
+            del db_write_keys
```
Comment (onewland) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:145]: is this just a safeguard against actually using this data wrong?

Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/postgres_v2.py:145]: yeah pretty much. it's the reason i tend to overwrite variables

**Pull request merged**